### PR TITLE
doc: update ivshmem user guide

### PR DESCRIPTION
**Pull request by yonghuah**
refine the guide to setup ivshmem for both hv-land and
  dm-land usage.

Signed-off-by: Yonghua Huang <yonghua.huang@intel.com>

### CODE REVIEW

File: doc/tutorials/enable_ivshmem.rst [hunk 1]
```diff
@@ -63,10 +39,37 @@ steps:
...
+Follow ``Enable Ivshmem Support`` and
+and add this line as an ``acrn-dm`` boot parameter::
+
+     -s slot,ivshmem,shm_name,shm_size
+
+where
+
+-  ``-s slot``  - Specify the virtual PCI slot number
+
+-  ``ivshmem``  - Virtual PCI device name
```
Comment (gvancuts) [inline doc/tutorials/enable_ivshmem.rst:56]: This phrasing makes me believe, as a user, that I may have to pick a name and use it there. So how about something like this instead:
```suggestion
-  ``ivshmem``  - Virtual PCI device emulating the Shared Memory
```

File: doc/tutorials/enable_ivshmem.rst [hunk 2]
```diff
@@ -63,10 +39,37 @@ steps:
...
+-  ``shm_size`` - Shared memory size of selected ``shm_name``.
+
+.. note:: This device can be used with real-time VM (RTVM) as well.
+
+.. _ivshmem-hv:
+
+Ivshmem HV-Land Usage
+*********************
+
+ Follow ``Enable Ivshmem Support`` to setup HV-Land Ivshmem support.
```
Comment (gvancuts) [inline doc/tutorials/enable_ivshmem.rst:71]: Same as above, use:
```
`Enable Ivshmem Support`_
```

**Review by gvancuts — CHANGES_REQUESTED** [review]
Would you also have some sample code to add to this so people can easily test their set-ups? That would be much helpful! ;-)

File: doc/tutorials/enable_ivshmem.rst [hunk 3]
```diff
@@ -9,50 +9,26 @@ solution or hv-land solution, according to the usage scenario needs.
...
-Ivshmem HV-Land Usage
-*********************
-
-The ``ivshmem`` hv-land solution is disabled by default in ACRN. You can enable
+The ``ivshmem`` solution is disabled by default in ACRN. You can enable
 it using the :ref:`ACRN configuration toolset <acrn_config_workflow>` with these
 steps:
 
-- Enable ``ivshmem`` hv-land in ACRN XML configuration file.
+- Enable ``ivshmem`` in ACRN XML configuration file.
```
[the remark this answers is not in the document]
Comment (yonghuah) [reply doc/tutorials/enable_ivshmem.rst:19]: make sense, will refine this desc. thanks.

File: doc/tutorials/enable_ivshmem.rst [hunk 4]
```diff
@@ -63,10 +39,37 @@ steps:
...
         communication between VM0 and VM2, it can be written as ``0:2``
 
-   .. note:: You can define up to eight ``ivshmem`` hv-land shared regions.
-
 - Build with the XML configuration, refer to :ref:`getting-started-building`.
 
+Ivshmem DM-Land Usage
+*********************
+
+Follow ``Enable Ivshmem Support`` and
```
[the remark this answers is not in the document]
Comment (yonghuah) [reply doc/tutorials/enable_ivshmem.rst:47]: this is what if want to do exactly, thank you @gvancuts

File: doc/tutorials/enable_ivshmem.rst (hunk 1 above)
Comment (yonghuah) [reply doc/tutorials/enable_ivshmem.rst:56]: make sense, will update.

File: doc/tutorials/enable_ivshmem.rst [hunk 5]
```diff
@@ -63,10 +39,37 @@ steps:
...
+     -s slot,ivshmem,shm_name,shm_size
+
+where
+
+-  ``-s slot``  - Specify the virtual PCI slot number
+
+-  ``ivshmem``  - Virtual PCI device name
+
+-  ``shm_name`` - Specify a shared memory name. This ``shm_name`` must be listed
+   in ``IVSHMEM_REGION`` in ``Enable Ivshmem Support`` section and needs to start
```
[the remark this answers is not in the document]
Comment (yonghuah) [reply doc/tutorials/enable_ivshmem.rst:59]: config tool will do sanity check on this and report error for this violation. @gvancuts

[the remark this answers is not in the document]
Comment (gvancuts) [reply doc/tutorials/enable_ivshmem.rst:59]: > 
> 
> config tool will do sanity check on this and report error for this violation. @gvancuts

Does the config tool check also inspects launch scripts? What if the user modifies his launch script manually later, what would happen?

[the remark this answers is not in the document]
Comment (yonghuah) [reply doc/tutorials/enable_ivshmem.rst:59]: actually no at this point.  2 methods to do this:
1) manually edit launch script file, in this way, user shall ensure these parameters match with that defined via config tool GUI, else ivshmem device init will fail when acrn-dm boot.
2) use tool to generate launch script file, in this case, config tool will do below validation actions(echo from @yuchuyang :)   ):
Here are the 3 parts of validation:
a.	If a <shm_region> is listed in launchscript.xml, the “IVSHMEM_ENABLED" of scenario.xml must set to ‘y’
b.	If a <shm_region> is listed in launchscript.xml, <region_name> and <region_size> (refers to step2 and step3) must be listed and matched in “IVSHMEM_REGION” of scenario.xml.
c.	If a <shm_region> of launchscript.xml is valid in step b, the listed <uos id="1"> of launchescript.xml must be listed in the <vmid_list> (refers to step2) of scenario.xml.

File: doc/tutorials/enable_ivshmem.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (yonghuah) [reply doc/tutorials/enable_ivshmem.rst]: @gvancuts understood your comment about test code,  it's not easy to add that in this PR, we do have a ivshmem test code, but there are many coding style issues inside, and is far from public,  internally used by QA and Dev. btw,  it's not difficult for user to get similar  test code from internet, that's applicable in most cases when ACRN implements the spec compatible  emulated device.

[the remark this answers is not in the document]
Comment (gvancuts) [reply doc/tutorials/enable_ivshmem.rst]: Let's keep this as an open for now (i.e. not address it in this PR), it's already tracked in #5939

File: doc/tutorials/enable_ivshmem.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (gvancuts) [reply doc/tutorials/enable_ivshmem.rst]: Thanks, it sounds like there is still a fair amount of manual coordination to be done by the user (i.e. the one developing his own final scenario).

File: doc/tutorials/enable_ivshmem.rst [hunk 6]
```diff
@@ -9,63 +9,81 @@ solution or hv-land solution, according to the usage scenario needs.
...
+Enable Ivshmem Support
+**********************
+
+The ``ivshmem`` solution is disabled by default in ACRN. You can enable
+it using the :ref:`ACRN configuration toolset <acrn_config_workflow>` with these
+steps:
+
+- Enable ``ivshmem`` via ACRN configuration tool GUI.
+
+   - Set ``IVSHMEM_ENABLED`` to ``y``
```
Comment (gvancuts) [inline doc/tutorials/enable_ivshmem.rst:21]: @dbkinder , there is no way to refer to the config option automatically (just like we do with APIs using Doxygen), is there?

**Review by gvancuts — APPROVED** [review]

File: doc/tutorials/enable_ivshmem.rst (hunk 6 above)
Comment (dbkinder) [reply doc/tutorials/enable_ivshmem.rst:21]: yes, there is:  ``:option:`hv.FEATURES.IVSHMEM.IVSHMEM_ENABLED` ``

Comment (dbkinder) [reply doc/tutorials/enable_ivshmem.rst]: I force-pushed an update to add references to the option documentation, and will approve and merge.

**Comment by dbkinder** [comment]
I somehow pull in a code change from another PR.  I'll close this PR and open a new with with just the ivshmem document change.

**Pull request closed without merging**